src/tools/perf/perftest_mad.c

```diff
@@ -0,0 +1,697 @@
+
+    ret = umad_send(fd, agent, umad, len, timeout, 0);
+    if (ret < 0) {
+        goto failure;
+    }
+    free(umad);
+    return UCS_OK;
+failure:
+    free(umad);
+    return UCS_ERR_IO_ERROR;
+}
+
+static ucs_status_t perftest_mad_send(perftest_mad_rte_group_t *rte_group,
+                                      void *buffer, size_t size)
+{
+    const struct iovec iovec = {
+        .iov_base = buffer,
+        .iov_len  = size,
+    };
+    return perftest_mad_sendv(rte_group, &iovec, 1);
+}
+
+static void *realloc_or_free(void *old, size_t size)
+{
+    void *ptr = realloc(old, size);
+    if (!ptr) {
+        free(old);
+    }
+    return ptr;
+}
+
+static const char ping_str[] = "ping";
+
+static ucs_status_t perftest_mad_ping(perftest_mad_rte_group_t *rte_group)
+{
+    const struct iovec iovec = {
+        .iov_base = (void*)ping_str,
+        .iov_len  = sizeof(ping_str),
+    };
+    return perftest_mad_sendv(rte_group, &iovec, 1);
+}
+
+static int perftest_mad_is_ping(void *buffer, size_t size)
+{
+    size_t ping_size = sizeof(ping_str);
+    if (size != ping_size) {
+        return 0;
+    }
+    return memcmp(buffer, ping_str, ping_size) == 0;
+}
+
+static ucs_status_t perftest_mad_recv(perftest_mad_rte_group_t *rte_group,
+                                      void *buffer, size_t *avail,
+                                      ib_portid_t *remote_port)
+{
+    int ret;
+    void *umad;
+    uint8_t *data;
+    int len; /* cannot use 'size_t' here */
+    size_t ref_len;
+    struct ib_user_mad *user_mad;
+
+    int timeout = 3 * 1000;
+    int fd      = mad_rpc_portid(rte_group->mad_port);
+
+    ref_len = *avail + IB_VENDOR_RANGE2_DATA_OFFS;
+    umad    = calloc(1, ref_len + umad_size());
+    if (!umad) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ucs_debug("MAD: receive: Allocated len:%zu", ref_len);
+retry:
+    if (ref_len > INT_MAX - umad_size()) {
+        free(umad);
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    user_mad = umad;
+    len      = ref_len;
+
+    ret = umad_recv(fd, umad, &len, timeout);
+    if (ret < 0) {
+        if (errno == ETIMEDOUT || errno == EAGAIN) {
+            free(umad);
+            return UCS_ERR_NO_PROGRESS;
+        }
+        if (errno != EINVAL && errno != ENOSPC) {
+            ucs_error("MAD: receive: Failed, umad len:%d, ret:%d, errno:%d",
+                      len, ret, errno);
+            free(umad);
+            return UCS_ERR_IO_ERROR;
+        }
+        /* EINVAL: no length info, ENOSPC: needed length is set */
+        if (errno == EINVAL) {
+            ref_len *= 2; /* 'len' has become invalid */
+        } else {
+            ref_len = len;
+        }
+        umad = realloc_or_free(umad, umad_size() + ref_len);
+        if (!umad) {
+            return UCS_ERR_NO_MEMORY;
+        }
+        goto retry;
+    }
+
+    if (perftest_mad_get_remote_port(umad, remote_port) != UCS_OK) {
+        free(umad);
+        return UCS_ERR_IO_ERROR;
```

Comment:
minor: can use some label for error cleanup instead (as it is needed to do `free(umad)` for each error)

Reply:
fixed